WIP - #20744 - Cluster Sharding with remember-entity enabled fails to recover after restart.

pbarron opened the pull request:
Initial approach for solving the herd effect when restarting or rebalance a cluster when remember-entity is enabled. The approach recovers the underlying actors at a certain defined rate rather than all a once. We've tested the approach and it considerably reduces the latency on the underlying storage medium (cassandra in this case).

Commit:
Add entity recovery rate to persistent shard

Commit:
Add tests for entity recovery rate

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ClusterShardingSettings.scala:
```diff
@@ -83,7 +84,9 @@ object ClusterShardingSettings {
     val waitingForStateTimeout:                       FiniteDuration,
-    val updatingStateTimeout:                         FiniteDuration)
+    val updatingStateTimeout:                         FiniteDuration,
+    val entityRecoveryRateInterval:                   FiniteDuration)
```
I think there is a bit more work needed here to make it binary backwards compatible, the old constructor must be added

pbarron replied:
Ya... good catch. Will fix in next push.

patriknw commented:
That's a very nice solution. Perhaps it is enough to fix this old ticket: https://github.com/akka/akka/issues/15797 ?

patriknw commented:
OK TO TEST

drewhk commented on akka-cluster-sharding/src/main/resources/reference.conf:
```diff
@@ -94,6 +94,10 @@ akka.cluster.sharding {
   # works only for state-store-mode = "ddata"
   updating-state-timeout = 5 s
 
+  # The interval beteween which remembered entities are restarted when initialing the shard
```
This section is a bit terse. Can you please expand this and explain what exactly will happen if set for a particular non-zero value?

Also, I would probably prefer the setting to be either "off" or a positive number. That means that the name must be somewhat changed so that "off" makes sense.

patriknw replied:
@pbarron could you update the parsing of this property to handle `off`. We have similar in other places.

drewhk commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:
```diff
@@ -344,11 +345,24 @@ private[akka] class PersistentShard(
+      state.entities.foldLeft(entityRecoveryRateInterval) {
+        case (interval, entityId) ⇒
+          import context.dispatcher
+          context.system.scheduler.scheduleOnce(interval, self, RestartEntity(entityId))
```
Hm, I would probably like to see batches here. Maybe it makes more sense to provide a rate instead of a period?

NB: With streams it would be as easy as
```scala
Source(state.entities).throttle(initialBurst, entityPerSecond).runForeach(self ! RestartEntity(_))
```
Unfortunately this project cannot rely on streams.

drewhk replied:
Without streams it probably could be done as:
```scala
state.entities.grouped(batchsize).foldLeft(interval){ ... self ! RestartEntities(entityBatch) }
```

pbarron replied:
Cool... will take this approach.

drewhk commented:
This is nice, I think with some more polish it would be a great addition.

Commit:
Refactor to use a strategy approach

pbarron commented:
Thanks for your feedback. I've refactored the solution to take a more extensible approach. In the process I think I've addressed the above comments. If not let me know. I'll add more units if you are in agreement with the approach. 

@patriknw this PR goes some way to fix #15797 However, from reading the ticket it seems there also needs to be an additional feedback loop/back pressure in controlling how fast the entities are recovered. The above solution does take that in account. I've refactor the approach to use strategies so it is possible write a new strategy to take more feedback into account when starting the entities.

johanandren commented:
Binary compatibility failure
```
[error]  * method this(scala.concurrent.duration.FiniteDuration,scala.concurrent.duration.FiniteDuration,Int,scala.concurrent.duration.FiniteDuration,scala.concurrent.duration.FiniteDuration,scala.concurrent.duration.FiniteDuration,scala.concurrent.duration.FiniteDuration,scala.concurrent.duration.FiniteDuration,Int,Int,Int,scala.concurrent.duration.FiniteDuration,scala.concurrent.duration.FiniteDuration)Unit in class akka.cluster.sharding.ClusterShardingSettings#TuningParameters does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.cluster.sharding.ClusterShardingSettings#TuningParameters.this")
```

Commit:
Fix binary incompatilbity in clusterShardingSettings

Commit:
Add unit tests for entity recovery strategies

pbarron commented:
I've just added a number of unit tests for the entity recover strategies. 

Are you happy accept the PR as it stands? 

Are there other changes you like to see included?

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:
```diff
@@ -388,4 +405,60 @@ private[akka] class PersistentShard(
+abstract class EntityRecoveryStrategy {
+  import ShardRegion.EntityId
+  import akka.actor.ActorContext
+  def recoverEntities(context: ActorContext, entities: Set[EntityId])
```
add explicit return type `: Unit`

Passing around `ActorContext` is a no-go. It's too easy to access it from wrong thread.
Why is it needed? I see sending `RestartEntities` to `self` and I see `scheduler`.

`RestartEntities` is something internal and could not be used by a user defined strategy anyway, so it might be the wrong api if user can't do what we can in those strategies.

The `EntityRecoveryStrategy` could take an optional `ActorSystem` constructor parameter to get access to scheduler and dispatcher things.

The API could perhaps be
```
def recoverEntities(entities: Set[EntityId]): Future[Set[EntityId]]
```

When the Future is completed we pipe the `RestartEntities` messages to the shard actor, i.e. the sending is not done from the strategy.

Another thought is that this whole pluggable strategy mechanism might be complete overkill. Instead we could provide 2 fixed implementations and implement them in the shard actor.

pbarron replied:
Thank you for the feedback. 

> Another thought is that this whole pluggable strategy mechanism might be complete overkill. Instead we could provide 2 fixed implementations and implement them in the shard actor.

Happy to do this... the reason for initially taking this approach was to make it more extensible so that perhaps an addition implementation could be done for #15797 at a later stage. I've done the refactor locally... it is a lot simpler.. will push if you think it's best approach. Let me know.

ktoso replied:
I think we can still implement #15797 if we keep the APIs internally extensible, but not "totally extensible" as Patrik suggested.

I also agree with the comment about ActorContext and the API returning a Future.
// Patrik is on vacation for a while now, I'll take over reviewing here :)

pbarron replied:
Lucky him! :-) 

Just we are on the same page here... There seems to be two possible solutions:
1.) Provide 2 fixed implementations and implement them in the shard actor with config that allows us to switch between them. That possibly being 

`entity-recovery-strategy = "<all|constant>"` 

2.) Keep the current implemented interfaces; apply Patrik suggestions (ActorContext and the API return a Future); remove the external pluggable mechanism in favour config stated in 1.

I think you are siding with option 2, ya?

ktoso replied:
Yes exactly. Let's not overcomplicate (explode the number of options people *could* set) for now. The two strategies make sense

pbarron replied:
Cool... I'll update the solution to follow option 2.

Commit:
Remove pluginable interface. Update strategy to return futures

pbarron commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:
```diff
@@ -405,60 +409,45 @@ private[akka] class PersistentShard(
-      context.self ! RestartEntities(entities)
+  import scala.concurrent.Future
+
+  def recoverEntities(entities: Set[EntityId]): Set[Future[Set[EntityId]]]
```
@ktoso had to change the interface a bit from what Patrik was suggesting so that a future could be returned for each scheduled increment of the recovery.

drewhk commented:
LGTM from my side.

johanandren commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ClusterShardingSettings.scala:
```diff
@@ -71,19 +74,61 @@ object ClusterShardingSettings {
+      entityRecoveryStrategy == "all" || entityRecoveryStrategy == "constant",
+      s"Unknown 'entity-recovery-strategy' [$entityRecoveryStrategy], valid values are 'all' or 'constant'")
+
+    def this(
```
This is for binary compatibility right? (If so please add a comment saying "for binary compatibility")

pbarron replied:
done.

johanandren commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ClusterShardingSettings.scala:
```diff
@@ -38,7 +38,10 @@ object ClusterShardingSettings {
       waitingForStateTimeout = config.getDuration("waiting-for-state-timeout", MILLISECONDS).millis,
-      updatingStateTimeout = config.getDuration("updating-state-timeout", MILLISECONDS).millis)
+      updatingStateTimeout = config.getDuration("updating-state-timeout", MILLISECONDS).millis,
+      entityRecoveryStrategy = config.getString("entity-recovery-strategy"),
```
I'd rather see this as captured with types (sealed trait enum ADT type of thing) so that unsupported values does not go by unnoticed and fails later down in the actor.

johanandren replied:
Maybe even mapping it to the two impl. classes right away to skip an extra indirection?

pbarron replied:
If you look at line 94 and 95 the field is validated so it shouldn't fail later on in the actor. This follows the same approach as is done for stateStoreMode in ClusterShardingSettings so keeping it consistent. Either way I'm happy to change this to use a more typed approach if you feel that this a better approach. 

In terms of just passing the class names... this was the approach taken in early revisions of this PR (i.e. a more pluggable setup) which was thought too much at this stage.

johanandren replied:
Yeah, I didn't mean go back to passing class names, rather putting the translation from name to type as early as possible. You decide if you think this is good enough!

johanandren commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:
```diff
@@ -388,4 +409,45 @@ private[akka] class PersistentShard(
+    entities.grouped(numberOfEntities).foldLeft((frequency, Set[Future[Set[EntityId]]]())) {
+      case ((interval, scheduledEntityIds), entityIds) ⇒
+        (interval + frequency, scheduledEntityIds + scheduleEntities(interval, entityIds))
+    }._2
```
Not sure if this matters: won't this cause a small storm of deadletter messages if the shard is stopped before all entities has started?

pbarron replied:
Ya, with this approach there is a possibility that if the shard is stopped that there will be deadletters. Not sure how we can get around that without the recovery logic being placed in the shard actor itself?

johanandren commented:
Some minor things and nitpicks commented. I think it looks good except for those.

Commit:
Adding comment to indicate constructor is include for binary compatibility

johanandren commented:
Thanks, my other two comments are maybe not so important, I think you should do something with the feedback from Endre and Patrik on the reference config as well.

Commit:
Remove redundate config

pbarron commented:
Ahh... my mistake those comments from Patrik and Endre were on config that is no longer required. It was from a previous iteration. The config is not remove. Sorry about that.

pbarron commented:
Seems that the build is stuck... it is possible to give it a nudge?

ktoso commented:
Yup, most certainly - and sorry that review here was a bit stalled, we'll get back here shortly! Vacation time struck us a bit...

ktoso commented:
PLS BUILD

pbarron commented:
Thanks. No worries.

johanandren commented:
LGTM!

johanandren commented:
Thanks @pbarron !

pbarron commented:
NP. Thanks for accepting it.

notxcain commented:
@pbarron thanks for your contribution! @johanandren @patriknw will this make its way to 2.4.9?

johanandren commented:
It was merged before 2.4.9RC1 so it should be in there. Noticed now there is no ticket referenced though.

notxcain commented:
Right, I checked locally against 2.4.8 by mistake.

notxcain commented:
Feature worth noticing!

ktoso commented:
Could you make a ticket for it @johanandren ?

pbarron commented:
Hi @ktoso, @johanandren 

The ticket can be found at  #20744

johanandren commented:
Yah, I got interrupted by meeting while looking for it, that's it. Sorry.
